test: improve test-async-hooks-http-parser-destroy

Trott opened the pull request:
Improve reporting in test-async-hooks-http-parser-destroy when failing.

Before, failures looked like this (edited slightly to conform to our
commit message 72-char line length restriction):

    The expression evaluated to a falsy value:
    assert.ok(destroyedIds.indexOf(createdAsyncId) >= 0)

Now, you get a slightly better idea of what's up. (Is it missing one ID?
More than one? All of them?):

    AssertionError [ERR_ASSERTION]: Expected values to be strictly deep-equal:
    + actual - expected ... Lines skipped

      [
        156,
    ...
        757,
    -   761,
        765
      ]

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

addaleax commented on test/parallel/test-async-hooks-http-parser-destroy.js:
```diff
@@ -59,3 +53,22 @@ server.listen(0, function() {
+process.on('SIGTERM', () => {
+  console.error('Received SIGTERM. (Timed out?)');
+  checkOnExit();
+  process.exit(1);
```
Do you need `checkOnExit()` here if you already call `process.exit()`?

Trott replied:
I suppose not!

BridgeAR commented on test/parallel/test-async-hooks-http-parser-destroy.js:
```diff
@@ -21,7 +21,12 @@ async_hooks.createHook({
+    if (createdIds.includes(asyncId)) {
+      // Work around https://github.com/nodejs/node/issues/26961.
+      // Remove the next line when #26961 is fixed.
+      if (!destroyedIds.includes(asyncId))
```
I would rather not include this part and instead mark this test as flaky. Otherwise we might forget about it and the test will silently work while it should fail.

The failure indicates that we have a bug somewhere that we should fix.

Trott replied:
I see what you're saying and I'm definitely sympathetic. The problem is that with the changes in this PR, the test will basically fail 100% of the time. And I'm generally not OK with marking a test "flaky" as a workaround for a test that fails on every run. But the change is (IMO) a genuine improvement. It makes the checks more thorough (which is why it increases the failure rate dramatically).

I guess the solution everyone can support would be to fix the underlying bug.

Trott replied:
Now that https://github.com/nodejs/node/commit/ece507394a563bd7e3555a840a3d6980ffbb01ac landed, let's see if that fixes the remaining issues here and if I can remove this workaround.... Going to run CI with this and then again without it....

Trott replied:
Looks like `HTTPPARSER` ids never show up now.  I imagine that's unintended....

Trott replied:
Oh, it totally was intentional. Guess I need to update the test to use HTTPINCOMINGMESSAGE or something like that. Or maybe the test can/should be removed entirely?

Trott replied:
Probably the thing to do is to simply not filter the asyncIDs at all. If V8 compiling ever finishes for me locally, I'll try that out....

Trott replied:
Alas, it did not appear to fix the bug. IDs show up in `destroy` without showing up in `create`. But I should at least update the test to use HTTPINCOMINGMESSAGE.

Trott commented:
Passes everywhere except when run in a worker (using `tools/test.py --worker`:

```console
not ok 186 parallel/test-async-hooks-http-parser-destroy
  ---
  duration_ms: 5.423
  severity: fail
  exitcode: 1
  stack: |-
    
    events.js:173
          throw er; // Unhandled 'error' event
          ^
    AssertionError [ERR_ASSERTION]: Expected values to be strictly deep-equal:
    + actual - expected ... Lines skipped
    
      [
        158,
    ...
        487,
    -   491,
    -   495,
    -   499,
    -   503,
    -   507
      ]
        at process.checkOnExit (/home/iojs/build/workspace/node-test-commit-custom-suites-freestyle/test/parallel/test-async-hooks-http-parser-destroy.js:61:10)
        at process.emit (events.js:201:15)
    Emitted 'error' event at:
        at Worker.[kOnErrorMessage] (internal/worker.js:161:10)
        at Worker.[kOnMessage] (internal/worker.js:171:37)
        at MessagePort.<anonymous> (internal/worker.js:104:57)
        at MessagePort.emit (events.js:196:13)
        at MessagePort.onmessage (internal/worker/io.js:68:8)
  ...
```

Trott commented:
Missing some `destroy` calls when run in a worker....

Trott commented:
Now that https://github.com/nodejs/node/issues/26961 has landed, this can be unblocked/unstalled. I've rebased, removed the workaround for https://github.com/nodejs/node/issues/26961, and I think this is ready to go? Re-reviews would be helpful. @addaleax @joyeecheung

Trott commented:
Argh! Still fails with `--worker`....

Trott commented:
OK, fixed for workers!

Trott commented:
Landed in 11e1e00
